packages/itwin/reports-config-widget/src/widget/components/BulkExtractor.ts

```diff
@@ -0,0 +1,186 @@
+
+export type ReportMappingType = CreateTypeFromInterface<ReportMapping>;
+
+export type ReportMappingAndMapping = ReportMappingType & {
+  mappingName: string;
+  mappingDescription: string;
+  iModelName: string;
+};
+
+export default class BulkExtractor {
+  private _reportRunIds = new Map<string, string[]>();
+  private _reportsClientApi: ReportsClient;
+  private _mappingsClientApi: MappingsClient;
+  private _iModelsClient: IModelsClient;
+  private _extractionClientApi: ExtractionClient;
+  private _accessToken: Promise<string>;
+  private _iModelNames = new Map<string, string>();
+
+  constructor(apiConfig: ReportsApiConfig) {
+    const url = generateUrl(REPORTING_BASE_PATH, apiConfig.baseUrl);
+    this._reportsClientApi = new ReportsClient(url);
+    this._mappingsClientApi = new MappingsClient(url);
+    this._extractionClientApi = new ExtractionClient(url);
+    this._iModelsClient = new IModelsClient({
+      api: { baseUrl: generateUrl(Constants.api.baseUrl, apiConfig.baseUrl) },
+    });
+    this._accessToken = apiConfig.getAccessToken();
+  }
+
+  public async getStates(reportIds: string[]): Promise<Map<string, ExtractionStates>> {
+    const stateByReportId = new Map<string, ExtractionStates>();
+    const stateByRunId = new Map<string, ExtractorState>();
+
+    for (const reportId of reportIds) {
+      const runs = this._reportRunIds.get(reportId);
+      if (!runs) {
+        stateByReportId.set(reportId, ExtractionStates.None);
+        continue;
+      }
+
+      const states: ExtractorState[] = [];
+      for (const runId of runs) {
+        const state = stateByRunId.get(runId);
+        if (state) {
+          states.push(state);
+        } else {
+          const runState = await this.getSingleState(runId, await this._accessToken);
+          states.push(runState);
+          stateByRunId.set(runId, runState);
+        }
+      }
+      const finalState = this.getFinalState(states);
+      stateByReportId.set(reportId, finalState);
+    }
+    return stateByReportId;
+  }
+
+  public clearJob(reportId: string) {
+    this._reportRunIds.delete(reportId);
+  }
+
+  private getFinalState(states: ExtractorState[]) {
+    if (states.includes(ExtractorState.Failed))
+      return ExtractionStates.Failed;
+
+    if (states.includes(ExtractorState.Queued))
+      return ExtractionStates.Queued;
+
+    if (states.includes(ExtractorState.Running))
+      return ExtractionStates.Running;
+
+    if (states.includes(ExtractorState.Succeeded))
+      return ExtractionStates.Succeeded;
+
+    return ExtractionStates.Failed;
+  }
+
+  private async getSingleState(runId: string, accessToken: string): Promise<ExtractorState> {
+    try {
+      const response = await this._extractionClientApi.getExtractionStatus(accessToken, runId);
+      return response.state;
+    } catch (error: any) {
+      handleError(error.status);
+    }
+    return ExtractorState.Failed;
+  }
+
+  public async startJobs(reportIds: string[]) {
```

Comment:
Create a separate method that would start extraction for a single reportId and would return an array of runIds if that report has data from multiple iModels.

Reply:
Separate method is no longer needed as `startJobs` ensures that multiple jobs for the same iModel would not be started.